cryptoki/src/functions/session_management.rs

```diff
@@ -28,8 +27,93 @@ impl Pkcs11 {
+
+    /// Opens a new management session on a token
+    ///
+    /// The purpose of a management session is to hold the state of all sessions opened on a token
+    /// by the running application.  This allows threads to open their own session, already at the
+    /// desired state set by the management session.
+    ///
+    /// # Arguments
+    ///
+    /// * `slot_id` - The slot to open the session on
+    /// * `flags` - The flags for the session
+    /// * `user_type` - The type of user to login
+    /// * `pin` - The PIN to use, or `None` if a separate authentication path exists
+    ///
+    /// # Errors
+    ///
+    /// If the session fails to open, the typical [`Error::Pkcs11`] will return.  If a management
+    /// session already exists for the token, then [`Error::ManagementSessionExists`] will be
+    /// returned.
+    pub fn open_management_session(
+        &self,
+        slot_id: Slot,
+        flags: SessionFlags,
+        user_type: UserType,
+        pin: Option<&str>,
+    ) -> Result<()> {
+        let mut map = self
+            .management_sessions
+            .lock()
+            .map_err(|_e| crate::Error::MutexPoisonError)?;
+        if let std::collections::hash_map::Entry::Vacant(_) = map.entry(slot_id) {
+            let session_handle = self.open_session(slot_id, flags)?;
+            let (pin, pin_len) = match pin {
+                Some(pin) => (pin.as_ptr() as *mut u8, pin.len()),
+                None => (std::ptr::null_mut(), 0),
+            };
+            unsafe {
+                Rv::from(get_pkcs11!(self, C_Login)(
+                    session_handle,
+                    user_type.into(),
+                    pin,
+                    pin_len.try_into()?,
+                ))
+                .into_result()
+                .map_err(|e| {
+                    // Close the session if we can't login and return the resulting error
+                    // Not using `get_pkcs11` here because the macro uses a '?' which is not
+                    // allowed in a closure that does not return a Result.
+                    // Justification for unwrap -> If C_CloseSession doesn't exist, the
+                    // library is frankly unusable.
+                    let _ = self.function_list.C_CloseSession.unwrap()(session_handle);
+                    e
+                })?;
+            }
+
+            let _ = map.insert(slot_id, session_handle);
+            Ok(())
+        } else {
+            Err(crate::Error::ManagementSessionExists)
+        }
+    }
 
-        Ok(Session::new(session_handle, self))
+    /// Closes the management session on the slot
+    ///
+    /// # Errors
+    ///
+    /// If the session fails to close, the typical [`Error::Pkcs11`] will return.  If a management
+    /// session does not exist on this token, then [`Error::ManagementSessionDoesNotExist`] will
+    /// be returned.
+    ///
+    /// Even if the close session call fails, the value will still be removed from the map
```

Comment:
What happens in that case? Does the app remain logged in?

Reply:
Possibly, yes.  The user could then create another session and call logout on that.

I don't think this is really that much of an issue as C_CloseSession only really fails if the session doesn't exist anymore or something really catastrophic occurred.